PUBDEV-4882: Expose nfolds argument in the AutoML R/Py clients

ledell opened the pull request:
We now expose the `nfolds` argument field to the user in AutoML.  Changes were made to AutoML Java backend and R and Py clients and docs.
- Added `nfolds` to R and Python AutoML interface with a default value of 5 (not auto-generated).
- Added runit and pyunit to test for `nfolds >=2` and `nfolds = 0`
- Added client-side checking for the `nfolds = 1` case with nice error messages.
- Added `nfolds` to AutoML user guide docs & made some other updates to cross-validation and nfolds user guid pages.

navdeep-G commented on h2o-automl/src/main/java/ai/h2o/automl/AutoML.java:
```diff
@@ -638,8 +638,13 @@ private void setCommonModelBuilderParams(Model.Parameters params) {
+        //params._nfolds = 5;
+        params._nfolds = buildSpec.build_control.nfolds;
+        if (buildSpec.build_control.nfolds > 1) {
+          // TO DO: below allow the user to specify this (vs Modulo)
```
These `TO-DO`'s should be JIRAs if they are not already.

ledell replied:
This should be fixed when we expose `fold_assignment` to the clients.  In this JIRA: https://0xdata.atlassian.net/browse/PUBDEV-5039

navdeep-G replied:
Great. Ty

navdeep-G commented on h2o-automl/src/main/java/ai/h2o/automl/AutoMLBuildSpec.java:
```diff
@@ -3,6 +3,7 @@
 import hex.ScoreKeeper;
 import hex.grid.HyperSpaceSearchCriteria;
 import hex.schemas.GridSearchSchema;
+//import hex.Model;
```
This comment is not needed

navdeep-G commented on h2o-automl/src/main/java/ai/h2o/automl/AutoMLBuildSpec.java:
```diff
@@ -49,6 +50,10 @@ public AutoMLBuildControl() {
+
+    // Cross-validation fold construction
+    public int nfolds = 5; 
+    //public Model.Parameters.FoldAssignmentScheme fold_assignment;
```
Any reason this comment is needed?

ledell replied:
Yeah I started working on both at the same time, but then decided to comment it out for now.

navdeep-G replied:
Cool, sounds good.

navdeep-G commented on h2o-py/h2o/automl/autoh2o.py:
```diff
@@ -72,6 +75,11 @@ def __init__(self,
+        #nfolds must not be negative or equal to 1:
+        assert nfolds >= 0, "nfolds cannot be negative. Use nfolds >=2 if you want cross-valiated metrics and Stacked Ensembles or use nfolds = 0 to disable."
+        assert nfolds is not 1, "nfolds = 1 is an invalid value. Use nfolds >=2 if you want cross-valiated metrics and Stacked Ensembles or use nfolds = 0 to disable."       
+
```
It seems `nfolds` just needs to be >=2. You can just report back what you got for `nfolds` in the exception so the user knows what was passed in.

rpeck replied:
Nice cleanup of some of the help strings in the @API annotations.  Might as well fix the others so they are also complete sentences.

Also, I feel like "nfolds == 1" should mean "don't use xval; it's only one fold", just like 0 does.  Thoughts?

ledell replied:
@rpeck Ok I can clean up the rest of the strings.  I think we should be consistent with all the H2O algos, which is that `nfolds == 1` is invalid.  Agreed that it would have made more sense in the beginning to use `nfolds = 1` instead of `nfolds = 0` as the "no CV" option.

navdeep-G replied:
Okay, then can check `assert nfolds >=2 or nfolds ==0` ? 

And yes, always good to report back.

navdeep-G commented on h2o-py/tests/testdir_algos/automl/pyunit_automl_args.py:
```diff
@@ -102,7 +102,27 @@ def prostate_automl():
+    print("Check nfolds = 0 works properly")
+    aml = H2OAutoML(project_name="aml_nfolds0", nfolds=0, max_models=3, seed=1)
+    aml.train(y="CAPSULE", training_frame=train)
+    # grab the last model in the leaderboard (which will not be an SE model) and verify that nfolds = 0
```
Is this always guaranteed?

rpeck replied:
Nit: _if nfolds is not 5:_ is not necessary; the type check is always valid to make.

ledell replied:
@navdeep-G It's guaranteed that there's no ensembles because no SEs can be trained if `nfolds = 0` as long as it was passed through correctly.  So I'm checking that `nfolds = 0` is correctly passed through.

navdeep-G commented on h2o-r/h2o-package/R/automl.R:
```diff
@@ -157,6 +159,15 @@ h2o.automl <- function(x, y, training_frame,
+  }
+  if (nfolds == 1) {
+    stop("nfolds = 1 is an invalid value. Use nfolds >=2 if you want cross-valiated metrics and Stacked Ensembles or use nfolds = 0 to disable.")
+  }
```
Same as my comment on Py API. Just need to check `nfolds` >=2 and report back what was passed into the function in the exception.

rpeck replied:
Yay for spaces! }notthis{

ledell replied:
@navdeep-G I think you're forgetting about the valid `nfolds = 0` option.  That's why I broke it up into two separate checks.

navdeep-G replied:
Agreed @rpeck. Its the simple things :)

navdeep-G replied:
@ledell Yes, but the check can just be done in one swoop, i.e., `nfolds >=2 or nfolds == 0`.

ledell commented:
@rpeck @navdeep-G I have addressed all your concerns in this latest commit.  Let me know what you think.

I have also modified the `pyunit_automl_args.py` test because all the AutoML runs (tests) were getting added to the same AutoML project and that was not the intention.  It was creating new project names each time, but re-using the same H2OAutoML object, so updating only the project name did not create a new project/leaderboard.  I also edited the docstrings for readability.

ledell commented:
@h2o-ops please test!

ledell commented:
@h2o-ops please test!

navdeep-G commented:
LGTM. Still some tests failing?

ledell commented:
@navdeep-G Not sure why, I think it was passing before.  I'll try to re-run.

michal-raska commented:
@ledell regarding `pyunit_PUBDEV_3567_merge_stall_large.py`: this test has been failing a few days ago, but it's been fixed and is green now, maybe rebase on master would help

ledell commented:
Thanks for the suggestion @michal-raska.  I tried rebasing with master, but that didn't fix the issue.  I think it's okay to merge since the test is unrelated to AutoML.
